### ISSUE TITLE
CWEs in code

### ISSUE THREAD

**Issue by AndreyAlifanov** [curl/curl#8675]
Code has many CWEs.
PVS-Studio found 16 CWEs of first priority.

I've tested version 7.79.1, but as I can see, version 7.82.0 has the same code with same CWEs.

CWE-476 
conncache.c, line 554 The 'connc->closure_handle' pointer was utilized before it was verified against nullptr.
connect.c, line 1374, 1380 The 'conn->tempaddr[0]' pointer was utilized before it was verified against nullptr.
lib/url.c, linrs 4102, 4116 The 'conn' pointer was utilized before it was verified against nullptr.

CWE-754
lib/formdata.c, line 446 Consider checking for misprints. It's possible that the 'contenttype' should be checked here.

CWE-571
lib/ftp.c, line 1241 Expression 'sa->sa_family == 2' is always true.
lib/http.c, line 4280 Expression 'conn->httpversion != 20' is always true.
lib/http.c, line 4283 Expression 'conn->httpversion < 20' is always true.
lib/socks_sspi.c, line 566 Expression 'sspi_w_token[1].cbBuffer != 1' is always true.
lib/krb5_sspi.c, line 312 Expression 'input_buf[1].cbBuffer != 4' is always true.
lib/vtls/schannel.c, line 1276 Expression 'sspi_status == ((HRESULT) 0x00000000L)' is always true.

CWE-570
lib/ftp.c, line 3872 Expression 'wildcard->state == CURLWC_CLEAN' is always false.
lib/vtls/schannel_verify.c, line 316 Expression 'result != CURLE_OK' is always false.

CWE-682
lib/mime.c, line 1319 A call of the 'memset' function will lead to underflow of the buffer 'mime->boundary'.
lib/socks_sspi.c, line 589 A call of the 'memcpy' function will lead to underflow of the buffer 'socksreq'.

CWE-789
lib/rtsp.c, lines 696, 711 The value of the 'rtp_dataleft' variable is checked after it was used. Perhaps there is a mistake in program logic.

CWE-628
lib/transfer.c, line 343 The 'memcpy' function doesn't copy the whole string. Use 'strcpy / strcpy_s' function to preserve terminal null.

I would be very appreciated if you fix these CWEs in next release.

**Comment by danielgustafsson** [comment]
Please check these and report back which are actual bugs, and which are merely opinions by a static analyzer. Looking at a few I don't see any bugs. Also, the line numbers are no longer valid which makes it hard to follow the bouncing ball from this report, please re-run with the latest Git master.

**Comment by danielgustafsson** [comment]
To provide an example what I mean with opinions rather than bugs, consider this report:

> CWE-628 lib/transfer.c, line 343 The 'memcpy' function doesn't copy the whole string. Use 'strcpy / strcpy_s' function to preserve terminal null.

The actual code in question (I think, line numbers are incorrect in the report) reads:
```
315       /* copy the prefix to the buffer, leaving out the NUL */
316       memcpy(data->req.upload_fromhere, hexbuffer, hexlen);
```
The boundaries of this `memcpy` are quite clearly intentional. This is easy to see from reading the code, which is why static analyzer output should be seen as a starting point for investigations and not as a copy/pasteable truth.

**Comment by AndreyAlifanov** [comment]
Line numbers are correct for version 7.79.1, as I wrote.
And I undrestand that analyzer may be wrong.

I'm not the author of this code and have no opportunity to investigate it. I only need more clean and robust code.

**Comment by danielgustafsson** [comment]
> Line numbers are correct for version 7.79.1, as I wrote.

I don't question that at all, but development is only performed on Git master so it's not helpful to developers.

> I'm not the author of this code and have no opportunity to investigate it. I only need more clean and robust code.

Then I recommend to try and find reproducible bugs related to these reports, if there are none then maybe the code is quite robust?

**Comment by bagder** [comment]
> I only need more clean and robust code.

If you *need* it, then why not put in the extra effort and provide reports that we can work with easier?

Running a tool on an older version and throwing the raw output at us is not helping much and it doesn't magically create "robust code". In fact, I consider the code to be pretty robust already no matter what PVS-studio says.

**Comment by AndreyAlifanov** [comment]
Ok, I understand you.
If I've got some time, I'll try to analyze latest version of curl.

**Comment by danielgustafsson** [comment]
Feel free to open a new issue when you have re-analyzed (and preferably also analyzed the analysis).

**Comment by bagder** [comment]
>  lib/ftp.c, line 3872 Expression 'wildcard->state == CURLWC_CLEAN' is always false.

This is wrong. It is not always false as can easily be verified by reading the code.
